Make the contributing instructions more friendly (#5295)
tikv/tikv: CONTRIBUTING.md

If you are targeting platforms other than x86_64 linux, you'll also need:

* [`llvm` and `clang`](http://releases.llvm.org/download.html) - Used to generate bindings for different platforms and build native libraries (required for grpcio, rocksdb).

### Getting the repository

```
git clone https://github.com/tikv/tikv.git
cd tikv
# Future instructions assume you are in this directory
```

### Configuring your Rust toolchain

`rustup` is the official toolchain manager for Rust, similar to `rvm` or `rbenv` from the Ruby world.

TiKV is pinned to a version of Rust using a `rust-toolchain` file. `rustup` and `cargo` will automatically use this file. We also use the `rustfmt` and `clippy` components, to install those:

```bash
rustup component add rustfmt
rustup component add clippy
```

### Building and testing

TiKV includes a `Makefile` with common workflows, you can also use `cargo`, as you would in many other Rust projects.

At this point, you can build TiKV:

```bash
make build
```

During interactive development, you may prefer using `cargo check`, which will parse, borrow check, and lint your code, but not actually compile it:

```bash

> cargo check --all

Reviewer: Does this still cover everything because of #5289 ?

Author: Yes, the issue there is just that the binaries are not in the main crate, they are still in a crate which is part of the workspace

```

`cargo check` is particularly handy alongside `cargo-watch`, which runs a command each time you change a file.

```bash
cargo install cargo-watch
cargo watch -s "cargo check -all"
```

When you're ready to test out your changes, use the `dev` task. It formats your codebase, builds with `clippy` enabled, and runs tests. This should run without failure before you create a PR. Unfortunately, some tests might fail intermittently and others don't pass on all platforms. If you're unsure, just ask!

```bash
make dev
```

You can run the test suite alone, or just run a specific test:

```bash
# Run the full suite
make test
# Run a specific test
cargo test $TESTNAME
```

TiKV follows the Rust community coding style. We use Rustfmt and [Clippy](https://github.com/Manishearth/rust-clippy) to automatically format and lint our code. Using these tools is checked in our CI. These are included as a part of the `make dev` task.  You can also run them alone:

```bash
# Run Rustfmt

> cargo fmt --all

Reviewer: Does this still cover everything due to #5289 ?

Author: Yes, in fact I don't think we even need the `--all`

### Configuration

Read our configuration guide to learn about various [configuration options](./docs/reference/configuration). There is also a [configuration template](./etc/config-template.toml).

## Contribution flow

This is a rough outline of what a contributor's workflow looks like:

- Create a Git branch from where you want to base your work. This is usually master.
- Write code, add test cases, and commit your work (see [message format](#format-of-the-commit-message)).
- Run tests and make sure all tests pass.
- Push your changes to a branch in your fork of the repository and submit a pull request.
- Your PR will be reviewed by two maintainers (who may request some changes).
- Our CI systems automatically test all the pull requests.

Thanks for your contributions!

> ### Finding something to work on.

Reviewer 2: ```suggestion
### Finding something to work on
```

# Contributing to TiKV

Thanks for your interest in contributing to TiKV! This document outlines some of the conventions on building, running, and testing TiKV, the development workflow, commit message formatting, contact points and other resources.

If you need any help or mentoring getting started, understanding the codebase, or making a PR (or anything else really), please ask on [Slack](https://join.slack.com/t/tikv-wg/shared_invite/enQtNTUyODE4ODU2MzI0LTgzZDQ3NzZlNDkzMGIyYjU1MTA0NzIwMjFjODFiZjA0YjFmYmQyOTZiNzNkNzg1N2U1MDdlZTIxNTU5NWNhNjk), [Discourse](https://forum.tikv.org/c/tikv), or [WeChat](./README.md#WeChat).

> 

Reviewer 2: ```suggestion
```
Can we just use one blank line between sections?

If you are targeting platforms other than x86_64 linux, you'll also need:

* [`llvm` and `clang`](http://releases.llvm.org/download.html) - Used to generate bindings for different platforms and build native libraries (required for grpcio, rocksdb).

### Getting the repository

```
git clone https://github.com/tikv/tikv.git
cd tikv
# Future instructions assume you are in this directory
```

### Configuring your Rust toolchain

`rustup` is the official toolchain manager for Rust, similar to `rvm` or `rbenv` from the Ruby world.

TiKV is pinned to a version of Rust using a `rust-toolchain` file. `rustup` and `cargo` will automatically use this file. We also use the `rustfmt` and `clippy` components, to install those:

```bash
rustup component add rustfmt
rustup component add clippy
```

### Building and testing

TiKV includes a `Makefile` with common workflows, you can also use `cargo`, as you would in many other Rust projects.

At this point, you can build TiKV:

```bash
make build
```

During interactive development, you may prefer using `cargo check`, which will parse, borrow check, and lint your code, but not actually compile it:

```bash
cargo check --all
```

> It is particularly handy alongside `cargo-watch`, which runs a command each time you change a file.

Author: ```suggestion
`cargo check` is particularly handy alongside `cargo-watch`, which runs a command each time you change a file.
```

```bash
cargo install cargo-watch
cargo watch -s "cargo check -all"
```

> When you're ready to test out your changes, use the `dev` task. It will format your codebase, build with `clippy` enabled, and run tests. This should run without failure before you create a PR. Unfortunately, some tests will fail intermittently and others don't pass on all platforms. If you're unsure, just ask!

Author: ```suggestion
When you're ready to test out your changes, use the `dev` task. It formats your codebase, builds with `clippy` enabled, and runs tests. This should run without failure before you create a PR. Unfortunately, some tests might fail intermittently and others don't pass on all platforms. If you're unsure, just ask!
```

```bash
make dev
```

> You can run the test suite alone, or just run a specific test:

Author: I don't think this change adds much (and makes the phrasing a little awkward) - "alone" here is in contrast to with formatting.

```

When building with make, cargo will automatically use [pipelined][p] compilation to increase the parallelism of the build. To turn on pipelining while using cargo directly, set `CARGO_BUILD_PIPELINING=true`.

[p]: https://internals.rust-lang.org/t/evaluating-pipelined-rustc-compilation/10199

## Running TiKV

To run TiKV as an actual key-value store, you will need to run it as a cluster (a cluster can have just one node, which is useful for testing). You can do this on a single machine or on multiple machines. You need to use [PD](https://github.com/pingcap/pd) to manage the cluster (even if there is just one node on a single machine). Instructions are in our [docs](docs/how-to/deploy/using-binary.md) (if you build TiKV from source, then you don't need to download the binary).

### Configuration

Read our configuration guide to learn about various [configuration options](./docs/reference/configuration). There is also a [configuration template](./etc/config-template.toml).

## Contribution flow

This is a rough outline of what a contributor's workflow looks like:

- Create a Git branch from where you want to base your work. This is usually master.

> - Write code, add test cases, and commit your work (see below for message format).

Reviewer 2: ```suggestion
- Write code, add test cases, and commit your work (see [message format](#format-of-the-commit-message)).
```

- Run tests and make sure all tests pass.
- Push your changes to a branch in your fork of the repository and submit a pull request.

> - Your PR will be reviewed by two maintainers (who may request some changes).

Reviewer 2: ```suggestion
- Your PR is required to be reviewed by at least two maintainers, who may request some changes.
```

Author: These changes make review sound like a barrier, rather than a helpful discussion, which is something I'm trying to avoid with this PR

Reviewer 2: Indeed, although it's stating the fact. If the review is initiated by maintainers, your original wording is fine (I first thought the contributor needs to assign the reviewers - but they might not know any). But we can remove the parenthesis.

<BLANK LINE>
<why this change was made>
<BLANK LINE>
Signed-off-by: <Name> <email address>
```

The first line is the subject and should be no longer than 50 characters, the other lines should be wrapped at 72 characters (see [this blog post](https://preslav.me/2015/02/21/what-s-with-the-50-72-rule/) for why).

If the change affects more than one subsystem, you can use comma to separate them like `util/codec,util/types:`.

If the change affects many subsystems, you can use ```*``` instead, like ```*:```.

The body of the commit message should describe why the change was made and at a high level, how the code works.

> ### Signing off the Commit

Reviewer 2: ```suggestion
### Signing off the commit
```
